wolfcrypt/src/memory.c

```diff
@@ -1133,3 +1133,141 @@ void __attribute__((no_instrument_function))
+                wolfcrypt_irq_fpu_states[i] = (union fpregs_state *)kzalloc(PAGE_SIZE /* sizeof(union fpregs_state) */, GFP_KERNEL);
+                if (! wolfcrypt_irq_fpu_states[i])
+                    break;
+                /* double-check that the allocation is 64-byte-aligned as needed for xsave. */
+                if ((unsigned long)wolfcrypt_irq_fpu_states[i] & 63UL) {
+                    pr_err("warning, allocation for wolfcrypt_irq_fpu_states was not properly aligned (%px).\n", wolfcrypt_irq_fpu_states[i]);
+                    kfree(wolfcrypt_irq_fpu_states[i]);
+                    wolfcrypt_irq_fpu_states[i] = 0;
+                    break;
+                }
+            }
+            if (i < nr_cpu_ids) {
+                pr_err("warning, only %u/%u allocations succeeded for wolfcrypt_irq_fpu_states.\n", i, nr_cpu_ids);
+                return MEMORY_E;
+            }
+        }
+        return 0;
+    }
+
+    void free_wolfcrypt_irq_fpu_states(void) {
+        if (wolfcrypt_irq_fpu_states) {
+            unsigned int i;
+            for (i=0; i<nr_cpu_ids; ++i) {
+                if (wolfcrypt_irq_fpu_states[i])
+                    kfree(wolfcrypt_irq_fpu_states[i]);
+            }
+            kfree(wolfcrypt_irq_fpu_states);
+            wolfcrypt_irq_fpu_states = 0;
+        }
+    }
+
+    __must_check int save_vector_registers_x86(void) {
+        preempt_disable();
+        if (! irq_fpu_usable()) {
+            if (am_in_hard_interrupt_handler()) {
+                int processor_id;
+
+                if (! wolfcrypt_irq_fpu_states) {
+                    static int warned_on_null_wolfcrypt_irq_fpu_states = 0;
+                    preempt_enable();
+                    if (! warned_on_null_wolfcrypt_irq_fpu_states) {
+                        warned_on_null_wolfcrypt_irq_fpu_states = 1;
+                        pr_err("save_vector_registers_x86 with null wolfcrypt_irq_fpu_states.\n");
+                    }
+                    return EFAULT;
+                }
+
+                processor_id = __smp_processor_id();
+
+                if (! wolfcrypt_irq_fpu_states[processor_id]) {
+                    static int warned_on_null_wolfcrypt_irq_fpu_states_processor_id = -1;
+                    preempt_enable();
+                    if (warned_on_null_wolfcrypt_irq_fpu_states_processor_id < processor_id) {
+                        warned_on_null_wolfcrypt_irq_fpu_states_processor_id = processor_id;
+                        pr_err("save_vector_registers_x86 for cpu id %d with null wolfcrypt_irq_fpu_states[id].\n", processor_id);
+                    }
+                    return EFAULT;
+                }
+
+                /* check for nested interrupts -- doesn't exist on x86, but make
+                 * sure, in case something changes.
+                 *
+                 * (see https://stackoverflow.com/questions/23324084/nested-interrupt-handling-in-arm)
```

Comment:
Would prefer not having a link to stack overflow in our code. Perhaps document the details or put the link in this PR description instead.

Reply:
done